tests/functional_tests/p2p.py

```diff
@@ -158,6 +158,41 @@ def test_p2p_reorg(self):
             loops -= 1
             assert loops >= 0
 
+        # create a transaction which both daemons have in their pool
+        dst = {'address': '42ey1afDFnn4886T7196doS9GPMzexD9gXpsZJDwVjeRVdFCSoHnv7KPbBeGpzJBzHRCAs9UxqeoyFQMYbqSWYTfJJQAWDm', 'amount': 1000000000000}
+        res = self.wallet.transfer([dst])
+        txid = res.tx_hash
+        time.sleep(5)
+        for daemon in [daemon2, daemon3]:
+            res = daemon.get_transaction_pool_hashes()
+            assert len(res.tx_hashes) == 1
+            assert res.tx_hashes[0] == txid
+
+        # disconnect nodes and mine on daemon2, mining alt blocks containing that mempool tx
+        daemon2.out_peers(0)
+        daemon3.out_peers(0)
+        daemon2.pop_blocks(2)
+        daemon2.generateblocks('42ey1afDFnn4886T7196doS9GPMzexD9gXpsZJDwVjeRVdFCSoHnv7KPbBeGpzJBzHRCAs9UxqeoyFQMYbqSWYTfJJQAWDm', 5)
+
+        # assert that the tx is in daemon2's blockchain now, instead of the mempool
+        res = daemon2.get_transactions([txid])
+        assert len(res.txs) == 1
+        tx_details = res.txs[0]
+        assert not tx_details.in_pool
```

Comment:
You're right about that check being wrong. However, even your proposed changes aren't conservative enough if you want to handle popped blocks: if the chain purely goes backwards in time (which only normally happens when `pop_blocks` is called), a transaction output with a custom `unlock_time` might actually UNLOCK. This is because `Blockchain::get_adjusted_time()` is not monotonic, so an output that is unlocked now may become locked again in a future block.

Reply:
Interesting, so this should be good:

`if(txd.last_failed_id == m_blockchain.get_block_id_by_height(txd.last_failed_height) && txd.last_failed_height == m_blockchain.get_current_blockchain_height()-1)`

Reply:
Yes that should be good. If we wanted to get incredibly pedantic, we would also have to check that the hard fork version is greater than or equal to `HF_VERSION_DETERMINISTIC_UNLOCK_TIME`, since your system's wall-time might also not be monotonic, and consensus validation of a tx with a ring containing an output with a UNIX `unlock_time` isn't necessarily deterministic.

Reply:
Your call

Reply:
We should go with `if(txd.last_failed_id == m_blockchain.get_block_id_by_height(txd.last_failed_height) && txd.last_failed_height == m_blockchain.get_current_blockchain_height()-1)` IMO, since that wall-time thing won't affect any future or past transactions, it's only a technicality.